File: module-spi/src/main/java/com/dashingqi/module/spi/MainActivity.java
package com.dashingqi.module.spi;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.view.View;
import android.widget.Button;

import com.dashingqi.module.spi.intyerface.Play;

import java.util.Iterator;
import java.util.ServiceLoader;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Button loadInterface = findViewById(R.id.load_interface);
        loadInterface.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                loadMethod();
            }
        });
    }

    /**
     * SPI技术
     * Service Provider Interface
     * 查找接口实现的服务 多模块开发的时候
     */
    private void loadMethod() {
        ServiceLoader<Play> load = ServiceLoader.load(Play.class);
        Iterator<Play> iterator = load.iterator();
        while (iterator.hasNext()) {
            iterator.next().play();
        }
    }
}

File: module-view/src/main/java/com/chiatai/module_view/MainActivity.kt
package com.chiatai.module_view

import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.Button
import com.chiatai.module_view.tablayout.TabLayoutActivity

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        findViewById<Button>(R.id.jumpToTabLayout).setOnClickListener {
            Intent(this,TabLayoutActivity::class.java).apply {
                startActivity(this)
            }
        }

    }
}